### PR TITLE
Fix Gradle 6.9.x support

### LINKED ISSUE #1096: Plugin doesn't work on Gradle 6.9.x
### Expected Behavior
Build succeeds

### Current Behavior
Build fails with exception
```
> Task :dockerCreateDockerfile
FAILURE: Build failed with an exception.

* What went wrong:
Receiver class com.bmuschko.gradle.docker.tasks.image.Dockerfile$CopyFile does not define or inherit an implementation of the resolved method 'abstract java.lang.Object getProperty(java.lang.String)' of interface groovy.lang.GroovyObject.
2 actionable tasks: 2 executed
```

### Steps to Reproduce (for bugs)

Checkout the smoke-tests pr, run smoke tests on gradle 6.9.x

### Your Environment

Java 11, Gradle 6.9.x

### PR DESCRIPTION
**Pull request by Sineaggi**
Fixes a compatibility issue with Gradle 6.9.x

Without this builds fail with
```
> Task :dockerCreateDockerfile
FAILURE: Build failed with an exception.

* What went wrong:
Receiver class com.bmuschko.gradle.docker.tasks.image.Dockerfile$CopyFile does not define or inherit an implementation of the resolved method 'abstract java.lang.Object getProperty(java.lang.String)' of interface groovy.lang.GroovyObject.
2 actionable tasks: 2 executed
```

closes #1096

### CODE REVIEW

**Review by Jolley71717 — APPROVED** [review]
LGTM.

**Comment by bmuschko** [comment]
I am not sure if that will actually fix the issue. It seems more of a Groovy-compatibility issue as described [here](https://discuss.gradle.org/t/my-plugin-cant-be-applied-when-plugin-is-built-with-gradle-7-1-1/40464). My guess is that it is a binary-compatibility issue. One more reason to turn the Groovy production source code to Java.

**Comment by Sineaggi** [comment]
@bmuschko Unfortunately I haven't seen any Gradle devs chime in wrt binary compatibility issues wrt groovy 2 and 3. Wish binary-compatibility would be something mentioned in their compatibility page. At present the [existing docs](https://docs.gradle.org/current/userguide/compatibility.html#groovy) are a little sparse on details.

FWIW I ran some of the existing functional tests against 6.9.2 via the method [michel-kraemer](https://github.com/michel-kraemer) talked about in https://github.com/bmuschko/gradle-docker-plugin/pull/1095 and found a few more places where this issue popped up. This fix is closer to whack-a-mole.

**Comment by bmuschko** [comment]
They basically say this [here](https://docs.gradle.org/current/userguide/compatibility.html#groovy):

> Gradle plugins written in Groovy must use Groovy 3.x for compatibility with Gradle and Groovy DSL build scripts.

**Comment by donbeave** [comment]
Maybe it's better to migrate `DockerExtension` to Java as well, not only `DockerRegistryCredentials`? I can do this in my PR, as I'm already touching `DockerExtension` in it: https://github.com/bmuschko/gradle-docker-plugin/pull/1111

**Comment by donbeave** [comment]
@bmuschko I can do it. Looks like not a lot of work for me. The question is how to do it without affecting other work, like this PR and  all ongoing work on this issue: https://github.com/bmuschko/gradle-docker-plugin/issues/1102?

I think I can use this PR as a baseline. What do you think?

**Comment by bmuschko** [comment]
@donbeave That'd be great. Let's finish the two outstanding PRs and then we can start the work on converting Groovy to Java. It'll be less of a hassle. I am planning to merge the two outstanding PR by EOW.

File: src/main/java/com/bmuschko/gradle/docker/internal/ConventionPluginHelper.java [hunk 1]
```diff
@@ -15,8 +18,18 @@ private ConventionPluginHelper() {
...
     public static SourceSetOutput getMainJavaSourceSetOutput(Project project) {
-        JavaPluginExtension javaPluginExtension = project.getExtensions().getByType(JavaPluginExtension.class);
-        return javaPluginExtension.getSourceSets().getByName(SourceSet.MAIN_SOURCE_SET_NAME).getOutput();
+        SourceSetContainer sourceSets = getMainJavaSourceSetContainer(project);
+        return sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME).getOutput();
+    }
+
+    private static SourceSetContainer getMainJavaSourceSetContainer(Project project) {
+        try {
+            JavaPluginExtension javaPluginExtension = project.getExtensions().getByType(JavaPluginExtension.class);
```
Comment (bmuschko) [inline src/main/java/com/bmuschko/gradle/docker/internal/ConventionPluginHelper.java:27]: This class was introduced in Gradle 4.10: https://docs.gradle.org/current/javadoc/org/gradle/api/plugins/JavaPluginExtension.html. There's no need to catch this case where the class doesn't exist. We do not support pre-4.10 Gradle versions anymore.

**Comment by bmuschko** [comment]
Overall the PR looks good. A functional test is failing. Can you look into it? You may also want to merge `master` back into your branch. I fixed a couple of tests.

**Comment by bmuschko** [comment]
@Sineaggi Is this still something you are planning to pursue? If you are not, let's close it.

**Comment by bmuschko** [comment]
I am going to close this PR for now. I'd be more than happy to reopen if you want to pick this back up.

**Pull request closed without merging**